src/Features/CSharp/Portable/CodeRefactorings/AddNamedArguments/CSharpAddNamedArgumentsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,209 @@
+
+using System.Composition;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.AddNamedArguments;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.AddNamedArguments
+{
+    [ExtensionOrder(After = PredefinedCodeRefactoringProviderNames.IntroduceVariable)]
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpAddNamedArgumentsCodeRefactoringProvider)), Shared]
+    internal class CSharpAddNamedArgumentsCodeRefactoringProvider : AbstractAddNamedArgumentsCodeRefactoringProvider<ArgumentSyntax>
+    {
+        protected override bool TryGetOrSynthesizeNamedArguments(SeparatedSyntaxList<ArgumentSyntax> arguments, ImmutableArray<IParameterSymbol> parameters, out ArgumentSyntax[] namedArguments, out bool hasLiteral)
+        {
+            hasLiteral = false;
+
+            // Early return when already fixed
+            if (arguments[0].NameColon != null)
+            {
+                namedArguments = default(ArgumentSyntax[]);
+                return false;
+            }
+
+            var result = ArrayBuilder<ArgumentSyntax>.GetInstance(arguments.Count);
+            for (int index = 0; index < arguments.Count; ++index)
+            {
+                var argument = arguments[index];
+                if (argument.NameColon != null)
+                {
+                    result.Add(argument);
+                }
+                else
+                {
+                    var parameter = parameters[index];
+                    var namedArgument = argument.WithNameColon(NameColon(parameter.Name));
+                    if (parameter.IsParams && arguments.Count - index > 1)
+                    {
+                        var arrayBuilder = ArrayBuilder<ExpressionSyntax>.GetInstance(arguments.Count - index);
+                        for (; index < arguments.Count; ++index)
+                        {
+                            arrayBuilder.Add(arguments[index].Expression);
+                        }
+
+                        var expressions = arrayBuilder.ToArrayAndFree();
+                        var initializer = InitializerExpression(SyntaxKind.ArrayInitializerExpression, SeparatedList(expressions));
+                        var arrayType = parameter.Type.GenerateTypeSyntax() as ArrayTypeSyntax;
+                        var arrayCreation = ArrayCreationExpression(arrayType, initializer);
+                        result.Add(namedArgument.WithExpression(arrayCreation).WithTriviaFrom(argument.Expression));
+                        break;
+                    }
+                    else
+                    {
+                        if (!hasLiteral && IsLiteral(argument))
```

Comment:
Except that `|=` will evaluate rhs anyways. :D That's kinda micro optimization but I'll use yours.